engine/access/rest/websockets/controller.go

```diff
@@ -4,100 +4,225 @@ import (
-	streamApi state_stream.API,
-	streamConfig backend.Config,
-	conn *websocket.Conn,
+	dataProviderFactory dp.DataProviderFactory,
+	conn WebsocketConnection,
 ) *Controller {
 	return &Controller{
 		logger:               logger.With().Str("component", "websocket-controller").Logger(),
 		config:               config,
 		conn:                 conn,
 		communicationChannel: make(chan interface{}), //TODO: should it be buffered chan?
+		errorChannel:         make(chan error, 1),    // Buffered error channel to hold one error.
 		dataProviders:        concurrentmap.New[uuid.UUID, dp.DataProvider](),
-		dataProvidersFactory: dp.NewDataProviderFactory(logger, streamApi, streamConfig),
+		dataProvidersFactory: dataProviderFactory,
 	}
 }
 
-// HandleConnection manages the WebSocket connection, adding context and error handling.
+// HandleConnection manages the lifecycle of a WebSocket connection,
+// including setup, message processing, and graceful shutdown.
+//
+// Parameters:
+// - ctx: The context for controlling cancellation and timeouts.
 func (c *Controller) HandleConnection(ctx context.Context) {
-	//TODO: configure the connection with ping-pong and deadlines
+	defer close(c.errorChannel)
+	// configuring the connection with appropriate read/write deadlines and handlers.
+	err := c.configureConnection()
+	if err != nil {
+		// TODO: add error handling here
+		c.logger.Error().Err(err).Msg("error configuring connection")
+		c.shutdownConnection()
+		return
+	}
+
 	//TODO: spin up a response limit tracker routine
-	go c.readMessagesFromClient(ctx)
-	c.writeMessagesToClient(ctx)
+
+	// for track all goroutines and error handling
+	var wg sync.WaitGroup
+
+	c.startProcess(&wg, ctx, c.readMessagesFromClient)
+	c.startProcess(&wg, ctx, c.keepalive)
+	c.startProcess(&wg, ctx, c.writeMessagesToClient)
+
+	// Wait for context cancellation or errors from goroutines.
+	select {
+	case err, ok := <-c.errorChannel:
+		if !ok {
+			c.logger.Error().Msg("error channel closed")
+			//TODO: add error handling here
+			return
+		}
+		c.logger.Error().Err(err).Msg("error detected in one of the goroutines")
+		//TODO: add error handling here
+		c.shutdownConnection()
+	case <-ctx.Done():
+		// Context canceled, shut down gracefully
+		c.shutdownConnection()
+	}
+
+	// Ensure all goroutines finish execution.
+	wg.Wait()
+}
+
+// startProcess is a helper function to start a goroutine for a given process
+// and ensure it is tracked via a sync.WaitGroup.
+//
+// Parameters:
+// - wg: The wait group to track goroutines.
+// - ctx: The context for cancellation.
+// - process: The function to run in a new goroutine.
+//
+// No errors are expected during normal operation.
+func (c *Controller) startProcess(wg *sync.WaitGroup, ctx context.Context, process func(context.Context) error) {
+	wg.Add(1)
+
+	go func() {
+		defer wg.Done()
+
+		err := process(ctx)
+		if err != nil {
+			// Check if shutdown has already been called, to avoid multiple shutdowns
+			if c.shutdown {
```

Comment:
This is a data race, isn't it? I'm thinking of the following situation:
- One of the processes crashes with some error without calling shutdown.
- So, we got to this code when we read `c.shutdown` on this line
- Simultaneously, another process called shutdown and touches `c.shutdown` concurrently.

If we need this, we have to use an atomic variable here. However, I don't understand why we need it. The idea of `sync.Once` is to not have a variable that tracks if sth was called once or more times but to just call the function

Reply:
Updated `c.shutdown` to `atomic bool`